src/operator/batch_norm-inl.h

```diff
@@ -329,6 +339,128 @@ class BatchNormProp : public OperatorProperty {
   BatchNormParam param_;
 };  // class BatchNormProp
 
+namespace batchnorm {
+
+#if defined(__CUDACC__)
+#define __bn_hostonly__        __host__
+#define __bn_hostdevinl__  __host__ __device__ __forceinline__
+#define __bn_localinline__ __forceinline__
+#else
+#define __bn_hostonly__
+#define __bn_hostdevinl__  inline
+#define __bn_localinline__ inline
+#endif
+
+template<typename DType>
+class BNTensor3 {
```

Comment:
I think I was referring to don't use mshadow's template evaluation features. Like Tensor x; x = 1;
Only using the data structure is fine.

But you can leave it like this for now.